File: C/14_Backtracking/BOJ_16936_나3곱2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
long long int* b, *arr;
int n, *visited;
int compare(const void* a, const void* b) {
	long long int o1 = *(long long int*)a;
	long long int o2 = *(long long int*)b;
	if (o1 > o2) return 1;
	else if (o1 < o2) return -1;
	else return 0;
}
void dfs(int idx, int dpth) {
	if (!visited[idx]) {
		visited[idx] = 1;
		arr[dpth] = b[idx];
		if (dpth == n - 1) {
			for (int i = 0; i < n; i++) {
				printf("%lld ", arr[i]);
			}
			exit(0);
		}
		if (arr[dpth] % 3 == 0) {
			for (int i = idx - 1; i >= 0; i--) {
				if (b[i] == arr[dpth] / 3) {
					dfs(i, dpth + 1);
				}
			}
		}
		for (int i = idx + 1; i < n; i++) {
			if (b[i] == arr[dpth] * 2) {
				dfs(i, dpth + 1);
			}
		}
		visited[idx] = 0;
	}
}
main() {
	scanf("%d", &n);
	b = (long long int*)malloc(sizeof(long long int) * n);
	arr = (long long int*)malloc(sizeof(long long int) * n);
	visited = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%lld", &b[i]);
	}
	qsort(b, n, sizeof(long long int), compare);
	for (int i = 0; i < n; i++) {
		memset(visited, 0, sizeof(int) * n);
		dfs(i, 0);
	}
}

File: JAVA/14_Backtracking/BOJ_16936_나3곱2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static long[] b, arr;
    static boolean[] visited;

    public static void dfs(int idx, int dpth) {
        if (!visited[idx]) {
            visited[idx] = true;
            arr[dpth] = b[idx];
            if (dpth == n - 1) {
                StringBuilder sb = new StringBuilder();
                for (int i = 0; i < n; i++) {
                    sb.append(arr[i]).append(" ");
                }
                System.out.println(sb);
                System.exit(0);
            }
            if (arr[dpth] % 3 == 0) {
                for (int i = idx - 1; i >= 0; i--) {
                    if (b[i] == arr[dpth] / 3) {
                        dfs(i, dpth + 1);
                    }
                }
            }
            for (int i = idx + 1; i < n; i++) {
                if (b[i] == arr[dpth] * 2) {
                    dfs(i, dpth + 1);
                }
            }
            visited[idx] = false;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        b = new long[n];
        arr = new long[n];
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < n; i++) {
            b[i] = Long.parseLong(st.nextToken());
        }
        Arrays.sort(b);
        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            dfs(i, 0);
            Arrays.fill(visited, false);
        }
    }
}